File: 202-happy-number/202-happy-number.cpp
class Solution {
public:
    int squareAndSum(int num){
        int sum = 0;
        
        while(num){
            sum+= pow(num%10, 2);
            num = num/10;
        }
        return sum;
    }
    
    bool isHappy(int n) {
        int fast = n, slow = n;
        
        do{
            slow = squareAndSum(slow);
            fast = squareAndSum(fast);
            fast = squareAndSum(fast);
        }while(slow != fast);
        
        return fast == 1;
    }
};

File: 202-happy-number/202-happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        
        def sumAndSquare(n):
            add = 0
            
            while n:
                add += (n%10)**2
                n = n//10
            return add
            
        slow = sumAndSquare(n)
        fast = sumAndSquare(sumAndSquare(n))
        
        while slow != fast:
            slow = sumAndSquare(slow)
            fast = sumAndSquare(fast)
            fast = sumAndSquare(fast)
        return fast == 1
        